### PR TITLE
Picture in Picture support, AutoLayout improvements

### PR DESCRIPTION
**Pull request by saagarjha**
This pull request adds support for Picture in Picture. I would advise against merging this as-is, since it still needs a bit of work. I just wanted to get your feedback on it.

A couple of things that need addressing:

* Crash when switching between Picture in Picture and "normal" too quickly using the keyboard shortcut (Command+Shift+P)
* Some way of alerting the user on the main window that the video is playing as Picture in Picture
* You tell me

### CODE REVIEW

**Comment by matthewdias** [comment]
Could probably either do it Safari style, with a label on the window indicating PIP mode, or maybe just hide the window entirely?

**Comment by saagarjha** [comment]
I think hiding the window could lead to user confusion–plus, the controls for things like scrubbing and volume control are on the control bar. I'm thinking of adding a `NSVisualEffects` view with a label on it ("This video is playing in Picture in Picture") to replace the video view.

**Comment by saagarjha** [comment]
@matthewdias I've added an `NSVisualEffectView`+label.

**Comment by lhc70000** [comment]
Sorry for reverting some changes on `develop`.

When updating your branch, it's better and cleaner to use git rebase rather than merge.

**Comment by saagarjha** [comment]
Sure, no problem. Thankfully `project.pbxproj` is plain text, and not a binary file!

**Comment by saagarjha** [comment]
@lhc70000 I've remedied the conflicts. Please give this a thorough check, since I was forced to solve quite a few merge conflicts by hand (in the XIBs no less), so I can't guarantee that everything functions as expected. Please try out the new features you've added in 0.0.4 and see if they work. Keeping up with IINA's changes is hard, so the sooner this gets merged the easier it is for me.

**Comment by lhc70000** [comment]
I'll take a look tomorrow (too busy today, sorry)

**Comment by saagarjha** [comment]
Sure, no problem.

**Comment by lhc70000** [comment]
Could you resolve the conflicts? (sorry, this is the last time)

**Comment by saagarjha** [comment]
@lhc70000 Done.

File: iina/PlayerCore.swift [hunk 1]
```diff
@@ -135,6 +136,7 @@ class PlayerCore: NSObject {
         mpvController.setFlag(MPVOption.PlaybackControl.pause, true)
         setSpeed(1)
       }
+      statusPaused = !statusPaused
```
Comment (lhc70000) [inline iina/PlayerCore.swift]: Seems that `statusPaused` is an unused field. Please delete it directly, and use `playerCore.info.isPaused` instead.

**Review by lhc70000 — CHANGES_REQUESTED** [review]

File: iina/MainWindowController.swift [hunk 2]
```diff
@@ -181,10 +186,21 @@ class MainWindowController: NSWindowController, NSWindowDelegate {
   @IBOutlet weak var rightArrowLabel: NSTextField!
   @IBOutlet weak var osdVisualEffectView: NSVisualEffectView!
   @IBOutlet weak var osd: NSTextField!
+	@IBOutlet weak var pipOverlayView: NSVisualEffectView!
```
Comment (saagarjha) [inline iina/MainWindowController.swift]: Interface Builder does that–I've been meaning to file a Radar.

File: iina/MainWindowController.swift [hunk 3]
```diff
@@ -482,7 +499,7 @@ class MainWindowController: NSWindowController, NSWindowDelegate {
     if obj == 0 {
       // main window
       isMouseInWindow = false
-      if controlBar.isDragging { return }
+      if controlBar.isDragging || isInPIP { return }
```
Comment (saagarjha) [inline iina/MainWindowController.swift]: Good idea; I've moved all the checks to `hideUI()`.

File: iina/MainWindowController.swift [hunk 4]
```diff
@@ -617,7 +638,8 @@ class MainWindowController: NSWindowController, NSWindowDelegate {
 
     // Set back the window appearance
     self.window!.appearance = NSAppearance(named: NSAppearanceNameVibrantLight);
-
+    
+    guard !isInPIP else { return }
```
Comment (saagarjha) [inline iina/MainWindowController.swift]: Not sure why I put it so high up; I've moved it down in the function.

File: iina/MainWindowController.swift [hunk 5]
```diff
@@ -906,7 +928,7 @@ class MainWindowController: NSWindowController, NSWindowDelegate {
 
     // if cropped, try get real window size
     if origWidth != currWidth || origHeight != currHeight {
-      let scale = origWidth == currWidth ? winFrame.width / currWidth : winFrame.height / currHeight
+      let scale = origWidth == currWidth ? winFrame.width / currWidth: winFrame.height / currHeight
```
Comment (saagarjha) [inline iina/MainWindowController.swift]: Yes.

File: iina/MainWindowController.swift [hunk 6]
```diff
@@ -1201,16 +1225,19 @@ class MainWindowController: NSWindowController, NSWindowDelegate {
       }
 
       if sender.intValue == 0 { // Released
-        if maxPressure == 1 { // Single click ended
+        if maxPressure == 1 &&
+          (speedValueIndex < speeds / 2 - 1 ||
+          Date().timeIntervalSince(lastClick) < minimumPressDuration) { // Single click ended, 2x speed
           speedValueIndex = oldIndex - 1
-        } else { // Force Touch ended
+        } else { // Force Touch or long press ended
```
Comment (saagarjha) [inline iina/MainWindowController.swift]: Sorry, I couldn't resist :)

File: iina/MenuController.swift [hunk 7]
```diff
@@ -139,8 +140,9 @@ class MenuController: NSObject, NSMenuDelegate {
 
     // -- screen
     fullScreen.action = #selector(MainWindowController.menuToggleFullScreen(_:))
+    pictureInPicture.action = #selector(MainWindowController.menuTogglePIP(_:))
     alwaysOnTop.action = #selector(MainWindowController.menuAlwaysOnTop(_:))
-
+
```
Comment (saagarjha) [inline iina/MenuController.swift]: Got it.

File: iina/PlayerCore.swift (hunk 1 above)
Comment (saagarjha) [inline iina/PlayerCore.swift]: Alright, it's deleted.

**Review by lhc70000 — APPROVED** [review]

**Comment by matthewdias** [comment]
@saagarjha @lhc70000 the aspect ratio gets messed up when exiting fullscreen.

- open vid in fullscreen with a display with a different aspect ration than the video
- switch to pip
- exit fullscreen

pip will get resized to aspect ratio of the display, and vid is stretched across it.

**Comment by matthewdias** [comment]
also the pip getting resized happens when you first enter pip while in fullscreen. it is letterboxed until you exit fullscreen, at which point it stretches.

**Comment by saagarjha** [comment]
I believe that this may be due to b2fba05633e1cb41f7a6f0b0a305404c77f42149. @lhc70000 would you mind reminding me on how resizing/scaling works?

**Comment by lhc70000** [comment]
@saagarjha It was handled manually prior to 0.0.4, but now the view always has the same size with window and let mpv to handle letterboxes in full screen, and the layer size is queried at realtime when drawing each frame. Aspect ratio in window mode is only locked by window's aspect ratio.

So `videoView.videoSize` is actually unused now. There may be two options:
- stop updating `videoSize` in full screen, or
- remove `videoSize` and get the actual size from `playerCore.info.videoWidth` and `.videoHeight`.

**Comment by lhc70000** [comment]
Just pushed a quick fix (get the actual size from `playerCore.info.videoWidth` and `.videoHeight`) as 0.0.5 will be released soon.

**Pull request merged**